pkg/executor/build.go

```diff
@@ -227,6 +227,11 @@ func (s *stageBuilder) optimize(compositeKey CompositeCache, cfg v1.Config) erro
 	if !s.opts.Cache {
 		return nil
 	}
+	var buildArgs = s.args.Clone()
+	// Restore build args back to their original values
```

Comment:
Why do we need to restore them?

Reply:
I'll try to to illustrate it with an example.

Take 
```
ARG ARG1=val1
ARG ARG2=val2
```
as an example. Going through those two stages `optimize` will generate the following cache keys:
```
'' -- starting empty
ARG ARG1=val1 results in roughly 'ARG ARG1=val1'
ARG ARG2=val2 results now in 'ARG1=val1 ARG ARG2=val2'
```
As you can see the `ARG1` will become an env variable and as such is part of the following cache keys (this is defined in the dockerfile spec that args become envs for the subsequent calls).

Now after `optimize()` the `s.args` really contains `ARG1` & `ARG2` (if you look further down in `optimize` the meta commands are actually executed and now change the state).

Now if we do not restore the build args to the state when optimize started, the actual build steps will start with `ARG1` & `ARG2` prepopulated. This has two issues, taking the above Dockerfiles the cache keys will now roughly look like this:
```
'ARG1=val1 ARG2=val2' -- not starting empty because build args already defined
ARG ARG1=val1 results in roughly 'ARG1=val1 ARG2=val2 ARG ARG1=val1'
ARG ARG2=val2 results now in 'ARG1=val1 ARG2=val2 ARG ARG2=val2'
```
as you can see the cache key is now completely different, so this cached layer will never be found (because optimize looked for a different layer due to the different cache key). -- This can be seen in https://github.com/GoogleContainerTools/kaniko/issues/1910#issuecomment-1027947684 ; during the actual build TEST is always set even if not executed yet -- which is also the root cause of https://github.com/GoogleContainerTools/kaniko/issues/1912

Reply:
Oh and we also might need similar handling for `ENV` statements.

Reply:
Nevermind, `ENV` is fine, this issue seems only to affect `ARG`

Reply:
@imjasonh, as I understand it, the `optimize` function actually executes the meta-only commands in order to properly determine the environment state for the purposes of generating a meaningful cache key ([build.go, line 283](https://github.com/apollo13/kaniko/blob/1dfda2bb812c0d87ae4b2d1372b261f3ef5ff685/pkg/executor/build.go#L283).

I assume this was leaving the `s.args` member in a dirty state for the rest of the execution of the `build` method, and therefore was the cause of the first problem fixed by this PR.

Reply:
Yes that is correct.